### PR TITLE
#4 -  Login + reset password

### LINKED ISSUE #4: Login + reset password
Register users should be able to login to application. Also they should have possibility to reset their password.

### PR DESCRIPTION
**Pull request by PrabuckiDominik**
This should close #4 .
Added: Register and login
Implemented:  Reset, forgot password and mail verification

### CODE REVIEW

File: app/Http/Requests/Auth/RegisterUserRequest.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array<string, ValidationRule|array|string>
+     */
+    public function rules(): array
+    {
+        return [
+            "email" => "required|string|email|max:255|unique:" . User::class,
+            "name" => "required|string",
```
Comment (EwelinaSkrzypacz) [inline app/Http/Requests/Auth/RegisterUserRequest.php:30]: And what will happens if user fill name in 256 characters?

**Review by EwelinaSkrzypacz — COMMENTED** [review]
1. I think that in this situation I should see validation error

image

2. I don't get reset password e-mail

3. When I'm on register page and I click "Register" I should see some validation messages - now I don't see anything

File: app/Http/Controllers/PasswordResetLinkController.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+use Illuminate\Support\Facades\Password;
+use Illuminate\Support\Str;
+use Inertia\Inertia;
+use Inertia\Response;
+
+class PasswordResetLinkController extends Controller
+{
+    public function create(): Response
+    {
+        return Inertia::render("Auth/Forgot-Password");
```
Comment (kamilpiech97) [inline app/Http/Controllers/PasswordResetLinkController.php:22]:
```suggestion
        return Inertia::render("Auth/ForgotPassword");
```
Should not be better?

File: app/Http/Controllers/PasswordResetLinkController.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        );
+
+        return $status === Password::RESET_LINK_SENT
+            ? back()->with(["status" => __($status)])
+            : back()->withErrors(["email" => __($status)]);
+    }
+
+    public function resetCreate(string $token): Response
+    {
+        return Inertia::render("Auth/Reset-Password", ["token" => $token]);
```
Comment (kamilpiech97) [inline app/Http/Controllers/PasswordResetLinkController.php:40]:
```suggestion
        return Inertia::render("Auth/ResetPassword", ["token" => $token]);
```
same here

File: app/Http/Requests/Auth/AuthenticateSessionRequest.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            "email" => [
+                "required" => "The email field is required.",
+                "max" => "Your email is too long. It must not be greater than 255 characters.",
+                "unique" => "The email has already been taken.",
+                "email" => "Your email is invalid.",
+            ],
+
+            "password" => [
+                "required" => "The password field is required.",
+            ],
```
Comment (kamilpiech97) [inline app/Http/Requests/Auth/AuthenticateSessionRequest.php:40]: Lets move that into validation translations file.

File: app/Http/Requests/Auth/ForgotPasswordRequest.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    public function messages(): array
+    {
+        return [
+            "email" => [
+                "required" => "The email field is required.",
+                "max" => "Your email is too long. It must not be greater than 255 characters.",
+                "email" => "Your email is invalid.",
+            ],
+        ];
```
Comment (kamilpiech97) [inline app/Http/Requests/Auth/ForgotPasswordRequest.php:35]: Same here

File: app/Http/Requests/Auth/RegisterUserRequest.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            ],
+            "password" => [
+                "required" => "The password field is required.",
+                "min" => "Your password is too short. It must be at least 8 characters.",
+            ],
+            "school_id" => [
+                "required" => "The school field is required.",
+                "exists" => "Your school is invalid. Check it again.",
+            ],
+        ];
```
Comment (kamilpiech97) [inline app/Http/Requests/Auth/RegisterUserRequest.php:57]: Same here

File: app/Http/Requests/Auth/ResetPasswordRequest.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+                "max" => "Your email is too long. It must not be greater than 255 characters.",
+                "email" => "Your email is invalid.",
+            ],
+            "password" => [
+                "required" => "The password field is required.",
+                "min" => "Your password is too short. It must be at least 8 characters.",
+                "confirmed" => "The password field confirmation does not match.",
+            ],
+        ];
+    }
```
Comment (kamilpiech97) [inline app/Http/Requests/Auth/ResetPasswordRequest.php:46]: Same here

File: app/Mail/RegistrationMail.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+    /**
+     * Get the attachments for the message.
+     *
+     * @return array<int, Attachment>
+     */
+    public function attachments(): array
+    {
+        return [];
+    }
```
Comment (kamilpiech97) [inline app/Mail/RegistrationMail.php:52]:
```suggestion
    public function envelope(): Envelope
    {
        return new Envelope(
            subject: "Registration Mail",
        );
    }

    public function content(): Content
    {
        return new Content(
            view: "emails.verify-email",
        );
    }
```
Lets remove not needed comments

File: resources/js/Pages/Auth/Verify-Email.vue
Comment (kamilpiech97) [inline resources/js/Pages/Auth/Verify-Email.vue:1]: Lets change file name into PascalCase

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

**Review by EwelinaSkrzypacz — COMMENTED** [review]
When I reset password and I fill email that doesn't exist in database I see error

image

I think that we should give information something like "If your e-mail is in our database, you will receive email with reset password" or something like that

File: .env.example [hunk 9]
```diff
@@ -5,7 +5,7 @@ APP_KEY=base64:sCsJw8z+d/4ymp0OvzSip2h4Vp2hZZhpV2uOxgTqP94=
 APP_DEBUG=true
 APP_URL=http://interns2024b.blumilk.localhost
 
-APP_LOCALE=en
+APP_LOCALE=pl
 APP_FALLBACK_LOCALE=en
 APP_FAKER_LOCALE=en_US
```
Comment (kamilpiech97) [inline .env.example:10]: why in .env file? not in config?

File: app/Http/Controllers/RegisterUserController.php [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+use Illuminate\Support\Facades\Hash;
+use Illuminate\Support\Facades\Redirect;
+
+class RegisterUserController extends Controller
+{
+    public function store(RegisterUserRequest $request): RedirectResponse
+    {
+        $userExists = User::query()->where("email", $request->email)->exists();
+        $user = new User($request->validated());
+        $user->password = Hash::make($request->password);
```
Comment (kamilpiech97) [inline app/Http/Controllers/RegisterUserController.php:21]: Lets move that into if

File: app/Http/Requests/Auth/AuthenticateSessionRequest.php [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    public function authorize(): bool
+    {
+        return true;
+    }
+
+    public function rules(): array
+    {
+        return [
+            "email" => "required|email|max:255",
+            "password" => "required|string",
```
Comment (kamilpiech97) [inline app/Http/Requests/Auth/AuthenticateSessionRequest.php:20]:
```suggestion
            "email" => ["required", "email", "max:255"],
            "password" =>  ["required", "string"],
```

File: app/Mail/RegistrationMail.php [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        );
+    }
+
+    /**
+     * @return array<int, Attachment>
+     */
+    public function attachments(): array
+    {
+        return [];
+    }
```
Comment (kamilpiech97) [inline app/Mail/RegistrationMail.php:41]:
```suggestion
```
is it needed?

File: app/Models/School.php [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+{
+    use HasFactory;
+
+    protected $fillable = [
+        "name",
+        "city",
+        "street",
+        "building_number",
+        "apartment_number",
+        "zipCode",
```
Comment (kamilpiech97) [inline app/Models/School.php:32]: Why one time `snake_case` and next `camelCase`?

File: database/migrations/2024_08_08_113859_create_schools_table.php [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    public function up(): void
+    {
+        Schema::create("schools", function (Blueprint $table): void {
+            $table->bigIncrements("id")->unique();
+            $table->string("name");
+            $table->string("city");
+            $table->string("street");
+            $table->string("building_number");
+            $table->string("apartment_number")->nullable();
+            $table->string("zipCode");
```
Comment (kamilpiech97) [inline database/migrations/2024_08_08_113859_create_schools_table.php:19]:
```suggestion
            $table->string("zip_code");
```

File: lang/pl/validation.php [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+        "surname" => [
+            "required" => "Pole :attribute jest wymagane.",
+            "max" => "Pole :attribute nie może być dłuższe niż :max znaków.",
+        ],
+        "name" => [
+            "required" => "Pole :attribute jest wymagane.",
+            "max" => "Pole :attribute nie może być dłuższe niż :max znaków.",
+        ],
+    ],
+
```
Comment (kamilpiech97) [inline lang/pl/validation.php:132]:
```suggestion
```

File: resources/js/Pages/Auth/ForgotPassword.vue [hunk 16]
```diff
@@ -0,0 +1,32 @@
+<script setup lang="ts">
+import { reactive } from 'vue'
+import { router } from '@inertiajs/vue3'
+
+const { errors, status } = defineProps<{
+  errors: Record<string, string[]>
+  status?: string
+}>()
+
+console.log(status)
```
Comment (kamilpiech97) [inline resources/js/Pages/Auth/ForgotPassword.vue:10]: ??

File: resources/js/Pages/Auth/ResetPassword.vue [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+
+const props = defineProps<
+{
+  token: string
+  errors: Record<string, string[]>
+}>()
+
+
+
```
Comment (kamilpiech97) [inline resources/js/Pages/Auth/ResetPassword.vue:13]:
```suggestion
```

File: resources/js/Pages/Auth/ResetPassword.vue [hunk 18]
```diff
@@ -0,0 +1,49 @@
+<script setup lang="ts">
+import { reactive } from 'vue'
+import { router } from '@inertiajs/vue3'
+
```
Comment (kamilpiech97) [inline resources/js/Pages/Auth/ResetPassword.vue:4]:
```suggestion
```

File: resources/js/Pages/Auth/ResetPassword.vue [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+    </div>
+    <div>
+      <label for="email">Password:</label>
+      <input v-model="form.password" name="password" type="password" required><br>
+      <label for="email">Password Confirmation:</label>
+      <input v-model="form.password_confirmation" name="password_confirmation" type="password"><br>
+      <div v-if="errors.password">{{ errors.password }}</div>
+    </div>
+
+
```
Comment (kamilpiech97) [inline resources/js/Pages/Auth/ResetPassword.vue:42]:
```suggestion
```

File: resources/js/Pages/Home.vue [hunk 20]
```diff
@@ -5,8 +5,16 @@ import AuthBanner from '@/components/Home/AuthBanner.vue'
...
 import BackgroundEffect from '@/components/Home/BackgroundEffect.vue'
+import type {School} from '@/Types/School'
 
 const isLogin = ref(null)
+
+const { errors, schools } = defineProps<{
+  errors: Record<string, string[]>
+  schools: School[]
+}>()
+
```
Comment (kamilpiech97) [inline resources/js/Pages/Home.vue:16]:
```suggestion
```

File: resources/js/components/Home/LoginForm.vue [hunk 21]
```diff
@@ -1,11 +1,32 @@
...
+
+const form = reactive({
+  email: null,
+  password: null,
+})
+
+function submit() {
+  router.post('/auth/login', form)
+}
+
```
Comment (kamilpiech97) [inline resources/js/components/Home/LoginForm.vue:17]:
```suggestion
```

File: routes/web.php [hunk 22]
```diff
@@ -2,15 +2,30 @@
...
+Route::post("/auth/register", [RegisterUserController::class, "store"])->name("register");
+Route::post("/auth/login", [AuthenticateSessionController::class, "authenticate"])->name("login");
+Route::get("/auth/logout", [AuthenticateSessionController::class, "logout"])->name("logout");
+Route::get("/email/verify", [EmailVerifyController::class, "create"])->middleware("auth")->name("verification.notice");
+Route::get("/email/{id}/{hash}", EmailVerifyController::class)->middleware(["signed", "throttle:6,1"])->name("verification.verify");
+Route::post("email/verification-notification", [EmailVerifyController::class, "send"])->middleware("auth", "throttle:6,1")->name("verification.send");
+
+Route::get("/auth/forgot-password", [PasswordResetLinkController::class, "create"])->middleware("guest")->name("password.request");
+Route::post("/auth/forgot-password", [PasswordResetLinkController::class, "store"])->middleware("guest")->name("password.email");
+Route::get("/auth/password/reset/{token}", [PasswordResetLinkController::class, "resetCreate"])->middleware("guest")->name("password.reset");
```
Comment (kamilpiech97) [inline routes/web.php:27]: Pack them into one group with `guest` middleware.

File: .env.example [hunk 23]
```diff
@@ -5,7 +5,6 @@ APP_KEY=base64:sCsJw8z+d/4ymp0OvzSip2h4Vp2hZZhpV2uOxgTqP94=
 APP_DEBUG=true
 APP_URL=http://interns2024b.blumilk.localhost
 
-APP_LOCALE=en
 APP_FALLBACK_LOCALE=en
 APP_FAKER_LOCALE=en_US
```
Comment (kamilpiech97) [inline .env.example:9]: That is needed too?

**Review by kamilpiech97 — APPROVED** [review]

File: app/Http/Requests/Auth/RegisterUserRequest.php (hunk 1 above)
Comment (PrabuckiDominik) [reply app/Http/Requests/Auth/RegisterUserRequest.php]: Now the user cannot pass a field containing more than 255 characters. Except for the password field.

**Review by kamilpiech97 — DISMISSED** [review]

**Review by AleksandraKozubal — APPROVED** [review]

**Review by kamilpiech97 — APPROVED** [review]

**Pull request merged**